File: three_sum/three_sum.go
package three_sum

import "sort"

func threeSum(nums []int) [][]int {
	ret := make([][]int, 0)
	sort.Ints(nums)
	length := len(nums)

	var (
		left  int
		right int
	)
	for i := 0; i < length; i++ {
		t := -nums[i]
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}
		left = i + 1
		right = length - 1
		for left < right {
			twoSum := nums[left] + nums[right]
			if twoSum < t {
				left++
			} else if twoSum > t {
				right--
			} else {
				ret = append(ret, []int{nums[i], nums[left], nums[right]})
				for left < right && nums[left+1] == nums[left] {
					left++
				}
				for left < right && nums[right] == nums[right-1] {
					right--
				}
				left++
				right--
			}
		}
	}
	return ret
}

File: three_sum/three_sum.ts
/**
 * @expect [[-1,0,1,2,-1,-4]] toBe [[[-1,-1,2],[-1,0,1]]]
 * @param nums
 */
function threeSum(nums: number[]): number[][] {
    const result: number[][] = [];
    const sorted = nums.sort((a, b) => a - b);
    let left, right;
    for (let i = 0, len = sorted.length; i < len; i++) {
        const target = -sorted[i];
        left = i + 1;
        right = len - 1;
        if (i > 0 && sorted[i] === sorted[i - 1]) {
            continue;
        }
        while (left < right) {
            if (sorted[left] + sorted[right] < target) {
                left++;
            } else if (sorted[left] + sorted[right] > target) {
                right--
            } else {
                result.push([sorted[i], sorted[left], sorted[right]])
                while (sorted[left + 1] === sorted[left]) left++;
                while (sorted[right - 1] === sorted[right]) right--;
                left++;
                right--;
            }
        }
    }
    return result
};
